src/gpuarray_array.c

```diff
@@ -338,44 +347,43 @@ static int gen_take1_kernel(GpuKernel *k, const gpuarray_buffer_ops *ops,
     atypes[apos++] = GA_SSIZE;
     atypes[apos++] = GA_SIZE;
   }
+  strb_appends(&sb, " GLOBAL_MEM const ga_ssize *ind, ga_size n0, ga_size n1,"
+               " GLOBAL_MEM int* err) {\n");
   atypes[apos++] = GA_BUFFER;
   atypes[apos++] = GA_SIZE;
   atypes[apos++] = GA_SIZE;
   atypes[apos++] = GA_BUFFER;
   assert(apos == nargs);
-  strb_appends(&sb, " GLOBAL_MEM const ga_size *ind, ga_size n0, ga_size n1,"
-               " GLOBAL_MEM int* err) {\n"
-               "  const ga_size idx0 = LDIM_0 * GID_0 + LID_0;\n"
-               "  const ga_size numThreads0 = LDIM_0 * GDIM_0;\n"
-               "  const ga_size idx1 = LDIM_1 * GID_1 + LID_1;\n"
-               "  const ga_size numThreads1 = LDIM_1 * GDIM_1;\n"
-               "  ga_size i0, i1;\n"
-               "  for (i0 = idx0; i0 < n0; i0 += numThreads0) {\n"
-               "    ga_ssize ii0 = ind[i0];\n"
-               "    ga_size pos0 = off;\n"
+  strb_appendf(&sb, "  const %s idx0 = LDIM_0 * GID_0 + LID_0;\n"
+               "  const %s numThreads0 = LDIM_0 * GDIM_0;\n"
+               "  const %s idx1 = LDIM_1 * GID_1 + LID_1;\n"
+               "  const %s numThreads1 = LDIM_1 * GDIM_1;\n"
+               "  %s i0, i1;\n", sz, sz, sz, sz, sz);
+  strb_appendf(&sb, "  for (i0 = idx0; i0 < n0; i0 += numThreads0) {\n"
+               "    %s ii0 = ind[i0];\n"
+               "    %s pos0 = off;\n"
                "    if (ii0 < 0) ii0 += d0;\n"
-               "    if ((ii0 < 0) || ((ga_size)ii0 >= d0)) {\n"
+               "    if ((ii0 < 0) || (ii0 >= d0)) {\n"
                "      *err = -1;\n"
                "      continue;\n"
                "    }\n"
-               "    pos0 += ii0 * s0;\n"
+               "    pos0 += ii0 * (%s)s0;\n"
                "    for (i1 = idx1; i1 < n1; i1 += numThreads1) {\n"
-               "      GLOBAL_MEM ga_byte *p = (GLOBAL_MEM ga_byte *)v;\n"
-               "      p += pos0;\n");
+               "      %s p = pos0;\n", ssz, sz, sz, ssz, sz);
   if (v->nd > 1) {
-    strb_appends(&sb, "      ga_size pos, ii = i1;\n");
+    strb_appendf(&sb, "      %s pos, ii = i1;\n", sz);
     for (i2 = v->nd; i2 > 1; i2--) {
       i = i2 - 1;
       if (i > 1)
-        strb_appendf(&sb, "      pos = ii %% d%u;\n"
```

Comment:
What the d in this line and the one bellow mean? Is there a special meaning for strb_appendf?